kernel/src/hil/time.rs

```diff
@@ -1,33 +1,270 @@
+    /// unsigned arithmetic and considering wraparound. It returns `true`
+    /// if, incrementing from `start`, the value will be reached before `end`.
+    /// Put another way, it returns `(self - start) < (end - start)` in
+    /// unsigned arithmetic.
+    fn within_range(self, start: Self, end: Self) -> bool;
+
+    /// Returns the maximum value of this type, which should be (2^width)-1.
+    fn max_value() -> Self;
+}
+
+/// Represents a clock's frequency in Hz, allowing code to transform
+/// between computer time units and wall clock time. It is typically
+/// an associated type for an implementation of the `Time` trait.
+pub trait Frequency {
+    /// Returns frequency in Hz.
+    fn frequency() -> u32;
+}
+
+/// Represents a moment in time, obtained by calling `now`.
+pub trait Time {
+    /// The number of ticks per second
     type Frequency: Frequency;
+    /// The width of a time value
+    type Ticks: Ticks;
 
-    /// Returns the current time in hardware clock units.
-    fn now(&self) -> W;
+    /// Returns a timestamp. Depending on the implementation of
+    /// Time, this could represent either a static timestamp or
+    /// a sample of a counter; if an implementation relies on
+    /// it being constant or changing it should use `Timestamp`
+    /// or `Counter`.
+    fn now(&self) -> Self::Ticks;
 
-    /// Returns the wrap-around value of the clock.
-    ///
-    /// The maximum value of the clock, at which `now` will wrap around. I.e., this should return
-    /// `core::u32::MAX` on a 32-bit-clock, or `(1 << 24) - 1` for a 24-bit clock.
-    fn max_tics(&self) -> W;
+    /// Returns the number of ticks in the provided number of seconds,
+    /// rounding down any fractions. If the value overflows Ticks it
+    /// returns `Ticks::max_value()`.
+    fn ticks_from_seconds(s: u32) -> Self::Ticks {
+        let val: u64 = Self::Frequency::frequency() as u64 * s as u64;
+        ticks_from_val(val)
+    }
+
+    /// Returns the number of ticks in the provided number of milliseconds,
+    /// rounding down any fractions. If the value overflows Ticks it
+    /// returns `Ticks::max_value()`.
+    fn ticks_from_ms(ms: u32) -> Self::Ticks {
+        let val: u64 = Self::Frequency::frequency() as u64 * ms as u64;
+        ticks_from_val(val / 1000)
+    }
+
+    /// Returns the number of ticks in the provided number of microseconds,
+    /// rounding down any fractions. If the value overflows Ticks it
+    /// returns `Ticks::max_value()`.
+    fn ticks_from_us(us: u32) -> Self::Ticks {
+        let val: u64 = Self::Frequency::frequency() as u64 * us as u64;
+        ticks_from_val(val / 1_000_000)
+    }
+}
+
+fn ticks_from_val<T: Ticks>(val: u64) -> T {
+    if val <= T::max_value().into_u32() as u64 {
+        T::from(val as u32)
+    } else {
+        T::max_value()
+    }
+}
+
+/// Represents a static moment in time, that does not change over
+/// repeated calls to `Time::now`.
+pub trait Timestamp: Time {}
```

Comment:
This trait is not implemented or used anywhere throughout this PR, is it actually needed?

Reply:
It seemed like a good idea.